src/stdlib/__tests__/parser.ts

```diff
@@ -0,0 +1,368 @@
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses arrow function expressions properly', () => {
+  const code = `
+    stringify(parse("x => x + 1;"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses arrow function assignments properly', () => {
+  const code = `
+    stringify(parse("const y = x => x + 1;"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses object notation', () => {
+  const code = `
+    stringify(parse("let x = {a: 5, b: 10, 'key': value};"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses property access', () => {
+  const code = `
+    stringify(parse("a[b]; a.b; a[5]; a['b'];"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses property assignment', () => {
+  const code = `
+    stringify(parse("a[b] = 5; a.b = value; a[5] = 'value'; a['b'] = 42;"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Parses loops', () => {
+  const code = `
+    stringify(parse("while (true) { continue; break; } for (let i = 0; i < 1; i = i + 1) { continue; break; } for (i = 0; i < 1; i = i + 1) { continue; break; }"), undefined, 2);
+  `
+  const context = mockContext(4)
+  const promise = runInContext(code, context, { scheduler: 'preemptive' })
+  return promise.then(obj => {
+    expect(JSON.stringify(context.errors, undefined, 2)).toBe('[]')
+    expect(obj).toMatchSnapshot()
+    expect(obj.status).toBe('finished')
+  })
+})
+
+test('Does not parse incomplete statements', () => {
```

Comment:
Should this be in the metacircular interpreter tests? It should already be checked in the `disallowed_syntax` tests. Same for the other syntax checks.

Reply:
True, the negative tests are a bit redundant. I'll remove the negative tests (but keep the positive ones).